File: pkg/logging/logging.go
// Package logging writes the checkin's log in the management-tool convention:
// one line per event as "[yyyy-MM-dd HH:mm:ss] LEVEL  message" appended to
// /Library/Managed Encryption/logs/crypt.log, rolled daily with thirty kept
// generations, owned by the tool rather than by a launchd redirect. The
// authorization plugin appends to the same file beside its unified-log entries.
package logging

import (
	"bytes"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
	"time"
)

const (
	// Dir is the logs directory of the Managed Encryption root.
	Dir = "/Library/Managed Encryption/logs"
	// File is the current log file inside Dir.
	File = "crypt.log"
	// Keep is how many rolled daily files are retained.
	Keep = 30
)

// Writer formats every line it receives with a timestamp and level and appends
// it to the log file; when stdout is a terminal the plain line is echoed there
// as well so an administrator running checkin by hand still sees the output.
type Writer struct {
	mu    sync.Mutex
	file  *os.File
	echo  bool
	level string
	now   func() time.Time
}

var std *Writer // nolint:gochecknoglobals

// Setup prepares the log directory, rolls yesterday's file, and routes the
// standard log package through the convention writer. It never fails the
// caller: a log that cannot be opened falls back to stderr.
func Setup() {
	w, err := Open(Dir, File, time.Now)
	if err != nil {
		fmt.Fprintf(os.Stderr, "crypt: cannot open %s: %v\n", filepath.Join(Dir, File), err)
		log.SetFlags(0)
		return
	}
	std = w
	log.SetFlags(0)
	log.SetOutput(w)
}

// Open creates dir, rolls the current file when it was last written on an
// earlier day, prunes rolled files beyond Keep, and opens the file for append.
func Open(dir, name string, now func() time.Time) (*Writer, error) {
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return nil, err
	}
	path := filepath.Join(dir, name)
	Roll(dir, name, now())
	f, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o644)
	if err != nil {
		return nil, err
	}
	echo := false
	if fi, statErr := os.Stdout.Stat(); statErr == nil && fi.Mode()&os.ModeCharDevice != 0 {
		echo = true
	}
	return &Writer{file: f, echo: echo, level: "INFO", now: now}, nil
}

// Roll renames dir/name to dir/<base>-<yyyy-MM-dd>.log when its last write
// happened before today, then removes the oldest rolled files beyond Keep.
func Roll(dir, name string, today time.Time) {
	path := filepath.Join(dir, name)
	fi, err := os.Stat(path)
	if err != nil {
		return
	}
	last := fi.ModTime()
	if sameDay(last, today) {
		return
	}
	base := strings.TrimSuffix(name, filepath.Ext(name))
	rolled := filepath.Join(dir, fmt.Sprintf("%s-%s.log", base, last.Format("2006-01-02")))
	if _, err := os.Stat(rolled); err == nil {
		// Two rolls on one day: keep the earlier file, append today's to it.
		if old, err := os.ReadFile(path); err == nil {
			if f, err := os.OpenFile(rolled, os.O_WRONLY|os.O_APPEND, 0o644); err == nil {
				_, _ = f.Write(old)
				_ = f.Close()
			}
		}
		_ = os.Remove(path)
	} else {
		_ = os.Rename(path, rolled)
	}
	prune(dir, base)
}

func prune(dir, base string) {
	matches, err := filepath.Glob(filepath.Join(dir, base+"-*.log"))
	if err != nil || len(matches) <= Keep {
		return
	}
	sort.Strings(matches) // yyyy-MM-dd names sort chronologically
	for _, stale := range matches[:len(matches)-Keep] {
		_ = os.Remove(stale)
	}
}

func sameDay(a, b time.Time) bool {
	ay, am, ad := a.Local().Date()
	by, bm, bd := b.Local().Date()
	return ay == by && am == bm && ad == bd
}

// Write implements io.Writer for the standard log package: every non-empty
// line becomes one INFO record.
func (w *Writer) Write(p []byte) (int, error) {
	for _, line := range bytes.Split(bytes.TrimRight(p, "\n"), []byte("\n")) {
		text := strings.TrimRight(string(line), "\r")
		if text == "" {
			continue
		}
		w.emit(w.level, text)
	}
	return len(p), nil
}

// Log writes one record at the given level (DEBUG, INFO, WARN, ERROR).
func (w *Writer) Log(level, format string, args ...interface{}) {
	w.emit(level, fmt.Sprintf(format, args...))
}

func (w *Writer) emit(level, text string) {
	w.mu.Lock()
	defer w.mu.Unlock()
	record := fmt.Sprintf("[%s] %-5s %s\n", w.now().Format("2006-01-02 15:04:05"), level, text)
	if w.file != nil {
		_, _ = w.file.WriteString(record)
	}
	if w.echo {
		fmt.Fprintln(os.Stdout, text)
	}
}

// Close releases the log file.
func (w *Writer) Close() error {
	if w.file == nil {
		return nil
	}
	return w.file.Close()
}

// Errorf records an ERROR line through the shared writer, or stderr before Setup.
func Errorf(format string, args ...interface{}) {
	if std == nil {
		fmt.Fprintf(os.Stderr, format+"\n", args...)
		return
	}
	std.Log("ERROR", format, args...)
}

// Warnf records a WARN line through the shared writer, or stderr before Setup.
func Warnf(format string, args ...interface{}) {
	if std == nil {
		fmt.Fprintf(os.Stderr, format+"\n", args...)
		return
	}
	std.Log("WARN", format, args...)
}

File: Sources/CryptCore/Logging.swift
/*
 Crypt

 Copyright 2025 The Crypt Project.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
import Foundation
import os.log

public let keychainLog = OSLog(subsystem: cryptBundleID, category: "Keychain")
public let filevaultLog = OSLog(subsystem: cryptBundleID, category: "Filevault")
public let prefLog = OSLog(subsystem: cryptBundleID, category: "Preferences")
public let enablementLog = OSLog(subsystem: cryptBundleID, category: "Enablement")
public let coreLog = OSLog(subsystem: cryptBundleID, category: "Core")
public let checkLog = OSLog(subsystem: cryptBundleID, category: "Check")
public let escrowLog = OSLog(subsystem: cryptBundleID, category: "Escrow")
public let authMechsLog = OSLog(subsystem: cryptBundleID, category: "AuthMechs")
public let serverLog = OSLog(subsystem: cryptBundleID, category: "Server")

// The management-tool logging convention: beside the unified log, every record
// is appended to /Library/Managed Encryption/logs/crypt.log as
// "[yyyy-MM-dd HH:mm:ss] LEVEL  Category: message". The authorization plugin
// and the checkin binary write the same file; checkin owns the daily roll.
// The plugin runs as root at the login window, so appending is always
// possible; a failure to write is ignored rather than allowed to interfere
// with authorization.
public let managedLogDirectory = "/Library/Managed Encryption/logs"
public let managedLogPath = managedLogDirectory + "/crypt.log"

/// How many rolled daily log files are kept.
public let managedLogGenerations = 30

public enum CryptLogLevel: Int, Sendable, Comparable {
  case debug = 0, info, warning, error

  public var label: String {
    switch self {
    case .debug: return "DEBUG"
    case .info: return "INFO"
    case .warning: return "WARN"
    case .error: return "ERROR"
    }
  }

  public static func < (lhs: CryptLogLevel, rhs: CryptLogLevel) -> Bool {
    lhs.rawValue < rhs.rawValue
  }
}

private let cryptLogCategories: [ObjectIdentifier: String] = [
  ObjectIdentifier(keychainLog): "Keychain",
  ObjectIdentifier(filevaultLog): "Filevault",
  ObjectIdentifier(prefLog): "Preferences",
  ObjectIdentifier(enablementLog): "Enablement",
  ObjectIdentifier(coreLog): "Core",
  ObjectIdentifier(checkLog): "Check",
  ObjectIdentifier(escrowLog): "Escrow",
  ObjectIdentifier(authMechsLog): "AuthMechs",
  ObjectIdentifier(serverLog): "Server",
]

private let managedLogQueue = DispatchQueue(label: cryptBundleID + ".managedlog")

private let managedLogStamp: DateFormatter = {
  let f = DateFormatter()
  f.locale = Locale(identifier: "en_US_POSIX")
  f.dateFormat = "yyyy-MM-dd HH:mm:ss"
  return f
}()

private let rolledLogStamp: DateFormatter = {
  let f = DateFormatter()
  f.locale = Locale(identifier: "en_US_POSIX")
  f.dateFormat = "yyyy-MM-dd"
  return f
}()

/// Runtime configuration of the managed log. `minimumLevel` drops quieter
/// records, and `echoToStandardOutput` mirrors them to the terminal so an
/// administrator running checkin by hand still sees the output.
public enum ManagedLog {
  private static let state = ManagedLogState()

  public static var minimumLevel: CryptLogLevel {
    get { state.minimumLevel }
    set { state.minimumLevel = newValue }
  }

  public static var echoToStandardOutput: Bool {
    get { state.echo }
    set { state.echo = newValue }
  }

  /// Prepares the log directory and rolls yesterday's file. Safe to call more
  /// than once; a directory that cannot be created is not fatal, the records
  /// simply go to the unified log and, when echoing, to stdout.
  public static func prepare(now: Date = Date()) {
    managedLogQueue.sync { roll(now: now) }
  }

  /// Appends one record, honouring `minimumLevel`.
  public static func write(_ level: CryptLogLevel, category: String, _ text: String) {
    guard level >= minimumLevel else { return }
    let record = "[\(managedLogStamp.string(from: Date()))] "
      + "\(level.label.padding(toLength: 5, withPad: " ", startingAt: 0)) \(category): \(text)\n"
    if echoToStandardOutput {
      FileHandle.standardOutput.write(Data(text.utf8) + Data("\n".utf8))
    }
    managedLogQueue.async { append(record) }
  }

  /// Renames the current log when it was last written on an earlier day and
  /// removes rolled files beyond `managedLogGenerations`.
  static func roll(directory: String = managedLogDirectory, name: String = "crypt.log", now: Date = Date()) {
    let fm = FileManager.default
    try? fm.createDirectory(atPath: directory, withIntermediateDirectories: true,
                            attributes: [.posixPermissions: 0o755])
    let path = directory + "/" + name
    guard let attrs = try? fm.attributesOfItem(atPath: path),
          let modified = attrs[.modificationDate] as? Date,
          !Calendar.current.isDate(modified, inSameDayAs: now)
    else { return }

    let base = (name as NSString).deletingPathExtension
    let rolled = "\(directory)/\(base)-\(rolledLogStamp.string(from: modified)).log"
    if fm.fileExists(atPath: rolled) {
      // Two rolls in one day: append the current file to the existing one.
      if let existing = FileHandle(forWritingAtPath: rolled), let old = fm.contents(atPath: path) {
        existing.seekToEndOfFile()
        existing.write(old)
        existing.closeFile()
      }
      try? fm.removeItem(atPath: path)
    } else {
      try? fm.moveItem(atPath: path, toPath: rolled)
    }
    prune(directory: directory, base: base)
  }

  private static func prune(directory: String, base: String) {
    let fm = FileManager.default
    guard let entries = try? fm.contentsOfDirectory(atPath: directory) else { return }
    let rolled = entries
      .filter { $0.hasPrefix(base + "-") && $0.hasSuffix(".log") }
      .sorted()  // yyyy-MM-dd names sort chronologically
    guard rolled.count > managedLogGenerations else { return }
    for stale in rolled.prefix(rolled.count - managedLogGenerations) {
      try? fm.removeItem(atPath: directory + "/" + stale)
    }
  }

  private static func append(_ record: String) {
    let fm = FileManager.default
    if !fm.fileExists(atPath: managedLogDirectory) {
      try? fm.createDirectory(atPath: managedLogDirectory, withIntermediateDirectories: true,
                              attributes: [.posixPermissions: 0o755])
    }
    if !fm.fileExists(atPath: managedLogPath) {
      fm.createFile(atPath: managedLogPath, contents: nil, attributes: [.posixPermissions: 0o644])
    }
    guard let handle = FileHandle(forWritingAtPath: managedLogPath) else { return }
    defer { handle.closeFile() }
    handle.seekToEndOfFile()
    handle.write(Data(record.utf8))
  }
}

/// Mutable configuration behind a lock, so the log can be reconfigured from the
/// command line while records are being written from any queue.
private final class ManagedLogState: @unchecked Sendable {
  private let lock = NSLock()
  private var level: CryptLogLevel = .info
  private var echoOut: Bool = isatty(STDOUT_FILENO) == 1

  var minimumLevel: CryptLogLevel {
    get { lock.withLock { level } }
    set { lock.withLock { level = newValue } }
  }

  var echo: Bool {
    get { lock.withLock { echoOut } }
    set { lock.withLock { echoOut = newValue } }
  }
}

private func managedLogLevel(_ type: OSLogType) -> CryptLogLevel {
  switch type {
  case .error, .fault: return .error
  case .debug: return .debug
  default: return .info
  }
}

/// Logs to the unified log exactly as os_log would, and appends the same
/// record to the managed log file. Format arguments are the os_log ones.
public func cryptLog(_ message: StaticString, log: OSLog = .default, type: OSLogType = .default, _ args: CVarArg...) {
  switch args.count {
  case 0: os_log(message, log: log, type: type)
  case 1: os_log(message, log: log, type: type, args[0])
  case 2: os_log(message, log: log, type: type, args[0], args[1])
  case 3: os_log(message, log: log, type: type, args[0], args[1], args[2])
  default: os_log(message, log: log, type: type, args[0], args[1], args[2], args[3])
  }
  let template = "\(message)"
    .replacingOccurrences(of: "%{public}", with: "%")
    .replacingOccurrences(of: "%{private}", with: "%")
  let text = args.isEmpty ? template : String(format: template, arguments: args)
  let category = cryptLogCategories[ObjectIdentifier(log)] ?? "Crypt"
  ManagedLog.write(managedLogLevel(type), category: category, text)
}

/// Records one line at the given level. The string form is what the checkin
/// binary uses; the plugin keeps the os_log-shaped `cryptLog` above.
public func cryptLog(_ level: CryptLogLevel, _ log: OSLog, _ text: String) {
  let osType: OSLogType
  switch level {
  case .debug: osType = .debug
  case .info: osType = .default
  case .warning: osType = .default
  case .error: osType = .error
  }
  os_log("%{public}@", log: log, type: osType, text)
  ManagedLog.write(level, category: cryptLogCategories[ObjectIdentifier(log)] ?? "Crypt", text)
}
